src/Microsoft.AspNet.Razor/Tokenizer/CSharpKeywordDetector.cs

```diff
@@ -88,7 +88,8 @@ internal static class CSharpKeywordDetector
             { "interface", CSharpKeyword.Interface },
             { "break", CSharpKeyword.Break },
             { "checked", CSharpKeyword.Checked },
-            { "namespace", CSharpKeyword.Namespace }
+            { "namespace", CSharpKeyword.Namespace },
+            { "when", CSharpKeyword.When }
```

Comment:
It's not functionally wrong, if you were to label it anything it's extra sensitive to provide an accurate user experience. 

**Explanation:** Today, keywords and identifiers are treated identically with one exception: the editing experience. Keywords have always been seen as pieces of C# that "may" affect the C# parsing mechanics of Razor where identifiers are just filler code. Edit handlers protect this behavior by ensuring the editor re-parses the document when it encounters tokens that are or begin with keywords as the user is typing (the alternative to this behavior is to not re-parse and to just let the C# compiler do the complaining).

That's where the issue comes in; for Razor we must ensure that we know the document structure so we know when to transition into or out of HTML. For the `when` case, it's definitely rare for a user to run into it. But i'd rather the editor ensure it's document state when it does run into the keyword rather than having malformed `try` blocks disrupt the Razor editing experience.

Now to comment on my previous notion of "making a bulk push instead of doing things piecemealed" we could eventually get keywords to behave more contextually to prevent unneeded reparses. However, to do this right would involve new infrastructure and removal or modification of the current bits that make the Razor editing experience what it is today.